cmdtesting/prompt.go

```diff
@@ -0,0 +1,220 @@
+	if p.failed {
+		return "", errors.New("prompter failed")
+	}
+	if len(p.ios) == 0 {
+		p.c.Errorf("unexpected prompt %q; expected none", text)
+		return "", errors.New("unexpected prompt")
+	}
+	if !p.c.Check(text, gc.Matches, p.ios[0].prompt) {
+		p.failed = true
+		return "", errors.Errorf("unexpected prompt %q; expected %q", text, p.ios[0].prompt)
+	}
+	reply := p.ios[0].reply
+	logger.Infof("prompt %q -> %q", text, reply)
+	p.ios = p.ios[1:]
+	return reply, nil
+}
+
+// CheckDone asserts that all the expected prompts
+// have been printed and all the replies read, and
+// reports whether the check succeeded.
+func (p *SeqPrompter) CheckDone() bool {
+	if p.failed {
+		// No point in doing the details checks if
+		// a prompt failed earlier - it just makes
+		// the resulting test failure noisy.
+		p.c.Errorf("prompter has failed")
+		return false
+	}
+	r := p.c.Check(p.ios, gc.HasLen, 0, gc.Commentf("unused prompts"))
+	r = p.c.Check(p.HasUnread(), gc.Equals, false, gc.Commentf("some input was not read")) && r
+	r = p.c.Check(p.Tail(), gc.Matches, p.finalText, gc.Commentf("final text mismatch")) && r
+	return r
+}
+
+// AssertDone is like CheckDone but aborts the test if
+// the check fails.
+func (p *SeqPrompter) AssertDone() {
+	if !p.CheckDone() {
+		p.c.FailNow()
+	}
+}
+
+// NewPrompter returns an io.ReadWriter implementation that calls the
+// given function every time Read is called after some text has been
+// written or if all the previously returned text has been read. The
+// function's argument contains all the text printed since the last
+// input. The function should return the text that the user is expected
+// to type, or an error to return from Read. If it returns an empty string,
+// and no error, it will return io.EOF instead.
+func NewPrompter(prompt func(string) (string, error)) *Prompter {
+	return &Prompter{
+		prompt: prompt,
+	}
+}
+
+// Prompter is designed to be used in a cmd.Context to
+// check interactive request-response sequences
+// using stdin and stdout.
+type Prompter struct {
+	prompt func(string) (string, error)
+
+	written      []byte
+	allWritten   []byte
+	pending      []byte
+	pendingError error
+}
+
+// Tail returns all the text written since the last prompt.
+func (p *Prompter) Tail() string {
+	return string(p.written)
+}
+
+// HasUnread reports whether any input
+// from the last prompt remains unread.
+func (p *Prompter) HasUnread() bool {
+	return len(p.pending) != 0
+}
+
+// Read implements io.Reader.
+func (p *Prompter) Read(buf []byte) (int, error) {
+	if len(p.pending) == 0 && p.pendingError == nil {
+		s, err := p.prompt(string(p.written))
+		if s == "" && err == nil {
+			err = io.EOF
+		}
+		p.written = nil
+		p.pending = []byte(s)
+		p.pendingError = err
+	}
+	if len(p.pending) > 0 {
+		n := copy(buf, p.pending)
+		p.pending = p.pending[n:]
+		return n, nil
+	}
+	if err := p.pendingError; err != nil {
+		p.pendingError = nil
+		return 0, err
+	}
+	panic("unreachable")
```

Comment:
Is this really necessary here? It looks to me like you've covered all the possibilities - when would you expect to panic here now?

If you strongly believe in its necessity, I think we'd want to unit test that we can reach it ;D